### ISSUE TITLE
Rule Proposal: prefer-optional-chaining

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#13430]
**Please describe what the rule should do:**

The new rule warns `obj && obj.prop`-like notations to suggest `obj?.prop`.

**What new ECMAScript feature does this rule relate to?**

Optional chaining

**What category of rule is this? (place an "X" next to just one item)**

[X] Suggests an alternate way of doing something (suggestion)

**Provide 2-3 code examples that this rule will warn about:**

```js
//✖ BAD
obj && obj.prop
obj != null ? obj.prop : undefined
obj && obj[key]
obj != null ? obj[key] : undefined
func && func()
func != null ? func() : undefined
if (func) func()
if (func) { func() }

//✔ GOOD
obj?.prop
obj?.[key]
func?.()
```

**Why should this rule be included in ESLint (instead of a plugin)?**

`obj && obj.prop`-like notation is a popular practice. This rule will help people to learn the new syntax Optional Chaining. This is about a language feature rather than third-party platforms or libraries.

**Are you willing to submit a pull request to implement this rule?**

Yes.

**Comment by aladdin-add** [comment]
friendly ping @mysticatea :
are you still working on this? it seems very useful! 👍

**Comment by bradzacher** [comment]
Commenting here to mention that it's not possible to have a rule which won't false positive.

```js
function foo(arg) {
  return arg && arg.foo;
}
```

This code looks like it should report by this lint rule, until you understand the types that flow through this function:
```js
foo(false);
foo({ foo: 'some string' });
```

This lint rule would suggest that someone change the return to `arg?.foo`. However this changes the return type of the function from `string | false` to `string | undefined` - which is potentially a breaking change (and is a completely breaking change for TS code).

Just adding this as a warning and a learning from [my implementation of this rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-optional-chain.md) from a year and a bit ago (https://github.com/typescript-eslint/typescript-eslint/pull/1213).

**Comment by flo-sch** [comment]
> However this changes the return type of the function from `string | false` to `string | undefined`

Strictly speaking, `string | false` is abusive in that case, the `foo` function you provides can already returns `undefined` in several cases.
(Perhaps this is because of a simplified example here, don't want to nitty-pick about it)
The type signature of that function should actually be `undefined | {actual type of arg.foo, might be "unknown" or anything unless "arg" is typed?}`.

```js
function foo(arg) {
  return arg && arg.foo;
}

foo() // undefined
foo({}) // undefined
foo({ foo: 'some string' }) // 'some string'
foo({ foo: 12345 }) // 12345
foo({ foo: false }) // false
foo(true) // undefined
foo(false) // false
```

However you are right about a potential false-positive in the case where "arg" is "false" here:

```js
function bar(arg) {
  return arg?.foo;
}

bar() // undefined
bar({}) // undefined
bar({ foo: 'some string' }) // 'some string'
bar({ foo: 12345 }) // 12345
bar({ foo: false }) // false
bar(true) // undefined
bar(false) // undefined --> difference with foo
```

**Comment by bradzacher** [comment]
I'd agree that the code definitely not good and that nobody should ever write it.

But people do, and people rely on this sort of behaviour.

https://github.com/typescript-eslint/typescript-eslint/issues/2963
https://github.com/typescript-eslint/typescript-eslint/issues/1893

And there are other problems as well, like

```
declare const foo: null | {value: string};

const value = foo && foo.value;
// typeof value === null | string

if (value === null) {
  // code here to handle null
}

// after rule fix

const value = foo?.value;
// typeof value === undefined | string

if (value === null) {
  // this code is now unreachable
} 
```
https://github.com/typescript-eslint/typescript-eslint/issues/1820

This code is much more reasonable, relying on `=== null` is a pretty common thing (though I always recommend `== null` as it handles both `null` and `undefined`, and IMO the two should always be handled the same).
Fixing this case will break code, and reporting here can be considered a false positive.

What I'm saying is that without type information and complete, omnipresent knowledge of the usage of the value - this rule will always have false positives, and a fixer is not always safe.

Suggestion fixers are an alternative, but from my experience users will just apply them without thinking, and will break their code.

I just wanted to drop my learnings to save you guys from making the same mistakes.

**Comment by flo-sch** [comment]
I completely agree with your vision, I just wanted to reflect on the fact that, in your previous example, `undefined` was already a potential value even without the optional chaining

But yeah, those use-case make it risky to include in a linter only IMO

There are probably some code-scanning tools that could be used in combination to handle those "problematic" cases?

**Comment by mdjermanovic** [comment]
By @bradzacher's examples, it seems that we cannot reliably know whether or not `obj && obj.prop` should be reported, so it might be better to avoid reporting this pattern. Even if we don't auto-fix the code and don't provide [suggestions](https://eslint.org/docs/developer-guide/working-with-rules#providing-suggestions), the linting error itself still suggests replacing this code with `obj?.prop`, which might be wrong and cause bugs.

We can definitely make a rule that reports only equivalent code like `obj != null ? obj.prop : undefined`, but it's questionable how useful that rule would be.

**Comment by pke** [comment]
how about this rule only applies when using together with && and not as explicit return values:

`obj && obj.foo && callSomeFunction(obj.foo)`

or JSX (where I wanted to use such rule and I opened a [discussion](https://github.com/eslint/eslint/discussions/15068#discussioncomment-1339359))

`obj && obj.foo && <Text>{obj.foo}</Text>`

**Comment by tajetaje** [comment]
> how about this rule only applies when using together with && and not as explicit return values:
> 
> `obj && obj.foo && callSomeFunction(obj.foo)`
> 
> or JSX (where I wanted to use such rule and I opened a [discussion](https://github.com/eslint/eslint/discussions/15068#discussioncomment-1339359))
> 
> `obj && obj.foo && <Text>{obj.foo}</Text>`

I really like this idea, except for how it creates inconsistent behavior across logically similar code (when adding/removing ` && callSomeFunction(obj.foo)`)

**Comment by kecrily** [comment]
This issue problem has been open for a long time, and the OP mentioned [the existence of such a rule in @typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-optional-chain.md).

Should we just close the issue or try to port the rule from @typescript-eslint?

**Comment by mdjermanovic** [comment]
It seems that this rule cannot work reliably without type information. The first versions of its equivalent in the typescript-eslint plugin (the `@typescript-eslint/prefer-optional-chain` rule) didn't use type information. There were too many false positives so the rule had to be updated to use type information (https://github.com/typescript-eslint/typescript-eslint/issues/4820).

Given the above, I'd suggest that we close this issue.

Another option could be to target a limited number of patterns where type information isn't needed, but I think such a rule would be of low value and wouldn't meet the expectations.

**Comment by nzakas** [comment]
There was a lot of interest in this rule so I'm hesitant to give up on it without a more thorough explanation. What are some examples that we couldn't catch without type information?

**Comment by mdjermanovic** [comment]
> What are some examples that we couldn't catch without type information?

The most common patterns that can be replaced with optional chaining, like `bar && bar.baz`, would have false positives whenever the expression before `?.` can evaluate to a falsy but not `undefined` value, e.g., `null` or `false`.

```js

const bar = null;

const foo1 = bar && bar.baz;
console.log(foo1); // `null`

const foo2 = bar?.baz;
console.log(foo2); // `undefined`
```

**Comment by nzakas** [comment]
I see what you mean. Still, I'm not sure that means we shouldn't pursue this rule. The likelihood that someone writes `bar && bar.baz` when `bar` is potentially `false` seems small to me. The `null` case is seems more likely but I'm also not sure if that's disqualifying.

From what I understand, the main concern is that changing to optional chaining may return a different value (`null` or `undefined`, or in the extreme, `false`). However, that value may not matter depending on usage. In all of the OP cases, the value doesn't matter because it's never being used: 

```js
//✖ BAD
obj != null ? obj.prop : undefined
obj != null ? obj[key] : undefined
func && func()
func != null ? func() : undefined
if (func) func()
if (func) { func() }
```

It's certainly safe to replace all of these. Here are some other patterns I see commonly that I think also fall into this category:

```js
const value = foo === undefined ? undefined : foo;

// from cli-engine.js
if (!(rule && rule.meta && rule.meta.deprecated)) {
	continue;
}

// from file-enumerator.js
if (stat && stat.isDirectory()) {
	return this._iterateFilesWithDirectory(absolutePath, isDot);
}

// from config.js
if (config.plugins && config.plugins[pluginName]) {
	// do something with config.plugins[pluginName]
}
```

We we want to avoid is direct assignment, like this:

```js
const value = foo && foo.bar;
```

In that case, it's not clear that change would be safe and so it shouldn't be flagged (although I'd argue that an option to check assignments would be welcome for those who want to be more aggressive -- I just wouldn't want to autofix these).

In my opinion, I think there are still plenty of cases that we can flag and autofix safely.

**Comment by bradzacher** [comment]
> The likelihood that someone writes bar && bar.baz when bar is potentially false seems small to me.

It's not as uncommon as you might expect -- which is why we rebuilt the typescript-eslint rule to use type information. It's surprisingly common for people to write code that has variables/properties whose value may be `false | T`.

Reporting in these instances is very, very frustrating for a user -- esp because there's no way for someone to reconfigure the rule to opt-out. They just have to add a disable comment at the obviously broken false-positive.

It's also any falsey value (`''` or `0`). For example one might write code like this:
```js
const x; // a potentially empty string
if (x && x.contains('some value')) { ... }
```
If the rule tells them to rewrite this to an optional chain then there's a minor perf change caused by the optional chain and also the loss of the short-circuit for the empty string case.

> From what I understand, the main concern is that changing to optional chaining may return a different value (null or undefined, or in the extreme, false). However, that value may not matter depending on usage

The problem is the "may" -- that value **_may_** matter.
At scale that "may" is a surprising number of cases and something as simple as changing from `null` to `undefined` can break code. Many codebases do not use `== null` and instead prefer the strict checks -- so this simple value change can easily break something downstream.

---

If you limit the rule just to cases where the value is discarded then yeah you can have a rule that flags changes that would not cause downstream value changes, but it does not account for the falsey-value cases where the rule would suggest changing to broken code.

We know *_for certain_* that people write code that would be broken by a non-type-aware rule from the bug reports that we got in typescript-eslint.

ESP if you're adding an autofixer to the rule -- that is bad. The golden rule of autofixers is *_do not break the build_*, so adding an autofixer that will knowingly break the build for some users would be the incorrect thing to do.

---

This is the journey we've already been on within typescript-eslint and is why we completely removed the non-type-aware.
We tried this approach, found it did not work.

**Comment by mdjermanovic** [comment]
> We we want to avoid is direct assignment, like this:
> 
> const value = foo && foo.bar;

I think it's not just assignments, but any code where the exact value is possibly used. For example, `return foo && foo.bar` or `func(foo && foo.bar)`.

> ```js
> func && func()
> if (func) func()
> if (func) { func() }
> ```

`func?.();` would be an incorrect replacement if `func` can be `false`.

**Comment by nzakas** [comment]
@bradzacher thanks for the details of what you've explored. That makes sense to me.

Given the updated info, I agree with @mdjermanovic that it doesn't make sense to pursue this rule.
